lib/manager/components/DeploymentViewer.js

```diff
@@ -665,3 +748,205 @@ class CustomConfig extends Component<{
+  }
+
+  render () {
+    const {
+      customFileEditIdx,
+      idx
+    } = this.props
+    const {
+      fileSource,
+      model: customFile
+    } = this.state
+    const isValid = customFile.contents || customFile.uri
+    const isEditing = idx === customFileEditIdx
+    const canEdit = customFileEditIdx === null
+
+    return (
+      <div className='custom-file'>
+        <ButtonGroup>
+          {isEditing && (
+            <Button
+              disabled={!isValid}
+              onClick={this._onSave}
+            >
+              Save
+            </Button>
+          )}
+          {isEditing && <Button onClick={this._onCancelEditing}>Cancel</Button>}
+          {canEdit && <Button onClick={this._onEdit}>Edit</Button>}
+          {canEdit && <Button onClick={this._onDelete}>Delete</Button>}
+        </ButtonGroup>
+        <FormControl
+          disabled={!isEditing}
+          onChange={this._onChangeFilename}
+          placeholder='Enter filename'
+          type='text'
+          value={customFile.filename}
+        />
+        <Checkbox
+          checked={customFile.useDuringBuild}
+          disabled={!isEditing}
+          onChange={this._onChangeBuildUse}
+        >
+          Use during graph build
+        </Checkbox>
+        <Checkbox
+          checked={customFile.useDuringServe}
+          disabled={!isEditing}
+          onChange={this._onChangeServeUse}
+        >
+          Use while running server
+        </Checkbox>
+        <FormGroup validationState={isValid ? null : 'error'}>
+          <ControlLabel>File source</ControlLabel>
+          <FormControl
+            componentClass='select'
+            disabled={!isEditing}
+            onChange={this._onChangeSource}
+            placeholder='select type'
+            style={{ marginBottom: 5 }}
+            value={fileSource}
+          >
+            <option value='raw'>From raw input</option>
+            <option value='uri'>Download from URI</option>
+          </FormControl>
+          {!isValid && <HelpBlock>Please set contents or uri!</HelpBlock>}
+          {fileSource === 'raw' && (
+            <FormControl
+              componentClass='textarea'
+              disabled={!isEditing}
+              style={{height: '125px'}}
+              placeholder='{"blah": true}'
+              onChange={this._onChangeContents}
+              value={customFile.contents}
+            />
+          )}
+          {fileSource === 'uri' && (
+            <span>
+              <FormControl
+                disabled={!isEditing}
+                placeholder='https://www.examle.com/file.json'
+                onChange={this._onChangeUri}
+                type='text'
+                value={customFile.uri}
+              />
+              {!customFile.uri && (
+                <HelpBlock>Enter either a HTTP(S) URL or AWS S3 URI</HelpBlock>
+              )}
+            </span>
+          )}
+        </FormGroup>
+      </div>
+    )
+  }
+}
```

Comment:
I think it's safe to say that this file is getting far too large! Could you please split it up a bit?